OpenRA.Mods.Common/Traits/Upgrades/UpgradeManager.cs

```diff
@@ -66,17 +85,31 @@ public UpgradeManager(ActorInitializer init)
 			});
 		}
 
-		public void GrantTimedUpgrade(Actor self, string upgrade, int duration)
+		public void GrantTimedUpgrade(Actor self, string upgrade, int duration, object source = null)
 		{
 			var timed = timedUpgrades.FirstOrDefault(u => u.Upgrade == upgrade);
 			if (timed == null)
 			{
-				timed = new TimedUpgrade(upgrade, duration);
+				timed = new TimedUpgrade(upgrade, duration, source);
 				timedUpgrades.Add(timed);
 				GrantUpgrade(self, upgrade, timed);
 			}
 			else
+			{
+				if (source != null)
+				{
+					var src = timed.Sources.FirstOrDefault(s => s.Source == source);
+					if (src == null)
+					{
+						timed.Sources.Add(new TimedUpgrade.UpgradeSource(duration, source));
```

Comment:
Hm... This would then break the convention that you call `AcceptsUpgrade` before you grant the upgrade. Thus making the `GrantTimeUpgrade` effectively a `TryGrantTimeUpgrade`. If `UpgradeMaxAcceptedLevel` shouldn't be broken for timed upgrades, it might make more sense to modify `AcceptsUpgrade` so that it takes into account the source...